File: bridge/AudioSetting.cpp
#include "./dart_api/dart_api.h"
#include "LabSound/LabSound.h"
#include "KeepNode.cpp"
using namespace lab;

DART_EXPORT const char * AudioSetting_name(int nodeId, int settingIndex) {
    auto setting = getKeepAudioSetting(nodeId, settingIndex);
    if(setting) {
        return setting->name().c_str();
    }
    return "";
}
DART_EXPORT const char * AudioSetting_shortName(int nodeId, int settingIndex) {
    auto setting = getKeepAudioSetting(nodeId, settingIndex);
    if(setting) {
        return setting->shortName().c_str();
    }
    return "";
}
DART_EXPORT int AudioSetting_type(int nodeId, int settingIndex) {
    auto setting = getKeepAudioSetting(nodeId, settingIndex);
    if(setting) {
        return static_cast<int>(setting->type());
    }
    return 0;
}
DART_EXPORT int AudioSetting_valueBool(int nodeId, int settingIndex) {
    auto setting = getKeepAudioSetting(nodeId, settingIndex);
    if(setting) {
        return setting->valueBool();
    }
    return 0;
}
DART_EXPORT float AudioSetting_valueFloat(int nodeId, int settingIndex) {
    auto setting = getKeepAudioSetting(nodeId, settingIndex);
    if(setting) {
        return setting->valueFloat();
    }
    return 0.0;
}
DART_EXPORT uint32_t AudioSetting_valueUint32(int nodeId, int settingIndex) {
    auto setting = getKeepAudioSetting(nodeId, settingIndex);
    if(setting) {
        return setting->valueUint32();
    }
    return 0;
}
DART_EXPORT int AudioSetting_valueBus(int nodeId, int settingIndex) {
    auto setting = getKeepAudioSetting(nodeId, settingIndex);
    if(setting) {
        return keepBus(setting->valueBus());
    }
    return -1;
}


DART_EXPORT void AudioSetting_setBool(int nodeId, int settingIndex, int v, int notify) {
    auto setting = getKeepAudioSetting(nodeId, settingIndex);
    if(setting) {
        setting->setBool(v > 0, notify > 0);
    }
}

DART_EXPORT void AudioSetting_setFloat(int nodeId, int settingIndex, float v, int notify) {
    auto setting = getKeepAudioSetting(nodeId, settingIndex);
    if(setting) {
        setting->setFloat(v, notify > 0);
    }
}

DART_EXPORT void AudioSetting_setUint32(int nodeId, int settingIndex, uint32_t v, int notify) {
    auto setting = getKeepAudioSetting(nodeId, settingIndex);
    if(setting) {
        setting->setUint32(v, notify > 0);
    }
}

DART_EXPORT void AudioSetting_setEnumeration(int nodeId, int settingIndex, int v, int notify) {
    auto setting = getKeepAudioSetting(nodeId, settingIndex);
    if(setting) {
        setting->setEnumeration(v, notify > 0);
    }
}

DART_EXPORT void AudioSetting_setString(int nodeId, int settingIndex, char const*const v, int notify) {
    auto setting = getKeepAudioSetting(nodeId, settingIndex);
    if(setting) {
        setting->setString(v , notify > 0);
    }
}

File: lab_sound_ffi/lib/src/lab_sound/audio_setting.dart

import 'package:lab_sound_ffi/lab_sound_ffi.dart';

import 'audio_context.dart';
import 'lab_sound.dart';
import '../extensions/ffi_string.dart';

enum AudioSettingType {
  None,
  Bool,
  Integer,
  Float,
  Enumeration,
  Bus
}

abstract class AudioSetting<T> {
  AudioContext ctx;
  final int settingId;
  final int nodeId;

  AudioSetting(this.ctx, this.nodeId, this.settingId);
  String get name => LabSound().AudioSetting_name(nodeId, settingId).toStr();
  String get shortName => LabSound().AudioSetting_shortName(nodeId, settingId).toStr();

  AudioSettingType get type => AudioSettingType.values[LabSound().AudioSetting_type(nodeId, settingId)];

  T get value;
  set value(T val) => setValue(val);
  setValue(T val, [bool notify = true]);

  @override
  String toString() => "AudioSetting<$shortName: $value>";
}

class AudioSettingBool extends AudioSetting<bool> {
  AudioSettingBool(AudioContext ctx, int nodeId, int settingId) : super(ctx, nodeId, settingId) {
    assert(this.type == AudioSettingType.Bool);
  }

  @override
  setValue(bool val, [bool notify = true]) => LabSound().AudioSetting_setBool(nodeId, settingId, val ? 1 : 0, notify ? 1 : 0);

  @override
  bool get value => LabSound().AudioSetting_valueBool(nodeId, settingId) > 0;
}
class AudioSettingInteger extends AudioSetting<int> {
  AudioSettingInteger(AudioContext ctx, int nodeId, int settingId) : super(ctx, nodeId, settingId) {
    assert(this.type == AudioSettingType.Integer);
  }

  @override
  setValue(int val, [bool notify = true]) => LabSound().AudioSetting_setUint32(nodeId, settingId, val, notify ? 1 : 0);

  @override
  int get value => LabSound().AudioSetting_valueUint32(nodeId, settingId);
}
class AudioSettingFloat extends AudioSetting<double> {
  AudioSettingFloat(AudioContext ctx, int nodeId, int settingId) : super(ctx, nodeId, settingId) {
    assert(this.type == AudioSettingType.Float);
  }

  @override
  setValue(double val, [bool notify = true]) => LabSound().AudioSetting_setFloat(nodeId, settingId, val, notify ? 1 : 0);

  @override
  double get value => LabSound().AudioSetting_valueFloat(nodeId, settingId);
}
class AudioSettingBus extends AudioSetting<AudioBus> {
  AudioSettingBus(AudioContext ctx, int nodeId, int settingId) : super(ctx, nodeId, settingId) {
    assert(this.type == AudioSettingType.Float);
  }

  @override
  setValue(AudioBus val, [bool notify = true]) => LabSound().AudioSetting_setBool(nodeId, settingId, val.resourceId, notify ? 1 : 0);

  @override
  AudioBus get value => AudioBus.fromId(LabSound().AudioSetting_valueBus(nodeId, settingId));
}


class AudioSettingEnumeration extends AudioSetting<int> {
  AudioSettingEnumeration(AudioContext ctx, int nodeId, int settingId) : super(ctx, nodeId, settingId) {
    assert(this.type == AudioSettingType.Integer);
  }

  @override
  setValue(int val, [bool notify = true]) => LabSound().AudioSetting_setEnumeration(nodeId, settingId, val, notify ? 1 : 0);

  @override
  int get value => LabSound().AudioSetting_valueUint32(nodeId, settingId);
}
